static/pane.ts

```diff
@@ -0,0 +1,195 @@
+
+    /**
+     * Emit analytics event for opening the pane tab. Typical implementation
+     * looks like this:
+     *
+     * ```ts
+     * ga.proxy('send', {
+     *   hitType: 'event',
+     *   eventCategory: 'OpenViewPane',
+     *   eventAction: 'RustMir',
+     * });
+     * ```
+     */
+    abstract registerOpeningAnalyticsEvent(): void
+
+    /** Optionally overridable code for initializing pane buttons */
+    registerButtons(state: BasePaneState /* typeof state */): void {}
+
+    /** Optionally overridable code for initializing event callbacks */
+    registerCallbacks(): void {}
+
+    abstract getPaneName(): string;
+    abstract onCompiler(id: number, compiler: any, options: any, editorId: number): void;
+    abstract onCompileResult(id: unknown, compiler: unknown, result: unknown): void;
+    abstract close(): void;
+
+    /** Initialize standard lifecycle hooks */
+    protected registerStandardCallbacks(): void {
+        this.fontScale.on('change', () => this.updateState());
+        this.container.on('destroy', () => this.close());
+        this.container.on('resize', () => this.resize());
+        this.eventHub.on('compileResult', (id, compiler, result) => {
+            this.onCompileResult(id, compiler, result);
+        });
+        this.eventHub.on('compiler', (id, compiler, options, editorId) => {
+            this.onCompiler(id, compiler, options, editorId);
+        });
+        this.eventHub.on('compilerClose', (id) => this.onCompilerClose(id));
+        this.eventHub.on('settingsChange', (settings) => this.onSettingsChange(settings));
+        this.eventHub.on('shown', () => this.resize());
+        this.eventHub.on('resize', () => this.resize());
+    }
+
+    protected setTitle() {
+        this.container.setTitle(this.getPaneName());
+    }
+
+    protected onCompilerClose(id: unknown) {
+        if (this.compilerInfo.compilerId === id) {
+            _.defer(() => this.container.close())
+        }
+    }
+
+    protected onDidChangeCursorSelection(event: monaco.editor.ICursorSelectionChangedEvent) {
+        if (this.isAwaitingInitialResults) {
+            this.selection = event.selection
+            this.updateState();
+        }
+    }
+
+    protected onSettingsChange(settings: SiteSettings) {
+        this.settings = settings;
+        this.editor.updateOptions({
+            contextmenu: settings.useCustomContextMenu,
+            minimap: {
+                enabled: settings.showMinimap,
+            },
+            fontFamily: settings.editorsFFont,
+            fontLigatures: settings.editorsFLigatures,
+        })
+    }
+
+    getCurrentState() {
+        const state = {
+            id: this.compilerInfo.compilerId,
+            editorId: this.compilerInfo.editorId,
+            selection: this.selection,
+        };
+        this.fontScale.addState(state);
+        return state;
+    }
+
+    updateState() {
+        this.container.setState(this.getCurrentState());
+    }
+
+    resize() {
```

Comment:
This got changed a few days ago. It now calls a function in `utils.ts` for most panes. Can that be set here, or will we need to override this in almost every pane?

Reply:
Done, could you double-check the implementation just to be sure? Thanks!